src_c/transform.c

```diff
@@ -570,6 +570,96 @@ scale_to(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj, int width,
     return retsurf;
 }
 
+SDL_Surface *
+grayscale(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj)
+{
+    SDL_Surface *src = pgSurface_AsSurface(srcobj);
+    SDL_Surface *newsurf;
+
+    if (!dstobj) {
+        newsurf = newsurf_fromsurf(src, srcobj->surf->w, srcobj->surf->h);
+        if (!newsurf)
+            return NULL;
+    }
+    else {
+        newsurf = pgSurface_AsSurface(dstobj);
+    }
+
+    if (newsurf->w != src->w || newsurf->h != src->h) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Destination surface must be the same size as source surface."));
+    }
+
+    if (src->format->BytesPerPixel != newsurf->format->BytesPerPixel) {
+        return (SDL_Surface *)(RAISE(
+            PyExc_ValueError,
+            "Source and destination surfaces need the same format."));
+    }
+
+    Uint32 *pixels = (Uint32 *)src->pixels;
+
+    SDL_LockSurface(newsurf);
+
+    int i;
+    for (i = 0; i < src->w * src->h; i++) {
+        Uint8 r, g, b, a;
+        SDL_GetRGBA(pixels[i], newsurf->format, &r, &g, &b, &a);
+
+        Uint32 grayscale_pixel = 0.212671 * r + 0.715160 * g + 0.072169 * b;
+
+        Uint32 transparent_pixel;
+        SDL_GetColorKey(src, &transparent_pixel);
+
+        int x = i % newsurf->w;
+        int y = i / newsurf->w;
+        Uint32 *const target_pixel =
+            (Uint32 *)((Uint8 *)newsurf->pixels + y * newsurf->pitch +
+                       x * newsurf->format->BytesPerPixel);
+        if (pixels[i] != transparent_pixel) {
+            Uint32 new_pixel = (0xFF << 24) | (grayscale_pixel << 16) |
+                               (grayscale_pixel << 8) | grayscale_pixel;
```

Comment:
This assumes that the target surface is laid out as ARGB or ABGR (alpha first in memory).

With a surface with the alpha channel in a different location it generates the wrong result--
![output_wrongchannels](https://user-images.githubusercontent.com/46412508/201453892-86e8d078-d5bd-4bbc-8c0e-fc790345128f.png)

The alpha info went into blue and the greyscale luminance went into the alpha channel.

To translate an image into that surface format, I used this code (I couldn't get the masks argument of pygame.Surface to work)--
```py
base_bytes = bytes(image.get_width() * 4 * image.get_height())
image_base = pygame.image.frombytes(base_bytes, image.get_size(), "ARGB")
image_base.blit(image, (0,0))
```

Reply:
Okay I have updated this to use SDL_MapRGBA